### PR TITLE
Update SavedTensorHooks TLS stack to use SafePyObject

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #131700

Previously, we must manually manage refcounting when updating the TLS saved variable stack. With this PR, things should be handled automatically by the SafePyObject.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Review by albanD — APPROVED** [review]
Looks great!

File: test/test_autograd.py [hunk 1]
```diff
@@ -555,6 +555,49 @@ def test_not_implemented_fwad(self):
...
+
+    #         x = torch.randn(3, 3, requires_grad=True)
+    #         with ctx_1:
+    #             x.sin().sum().backward()
+    #         return weakref.ref(unpack)
+
+    #     with disable_gc():
+    #         unpack_hook_ref = scope()
+    #         self.assertIsNone(unpack_hook_ref())
+
```
Comment (soulitzer) [inline test/test_autograd.py:600]: Temporarily commenting this out to see if there is some interaction with this test and the failing compiled autograd test
```
2024-07-29T21:37:39.6996815Z [31m[1m_________________ TestCompiledAutograd.test_autograd_cpp_node __________________[0m
2024-07-29T21:37:39.6997323Z Traceback (most recent call last):
2024-07-29T21:37:39.6997977Z   File "/var/lib/jenkins/workspace/test/inductor/test_compiled_autograd.py", line 1451, in test_autograd_cpp_node
2024-07-29T21:37:39.6998644Z     self.check_output_and_recompiles(fn, 2)
2024-07-29T21:37:39.6999327Z   File "/var/lib/jenkins/workspace/test/inductor/test_compiled_autograd.py", line 86, in check_output_and_recompiles
2024-07-29T21:37:39.6999998Z     actual = list(opt_fn())
2024-07-29T21:37:39.7000516Z   File "/var/lib/jenkins/workspace/test/inductor/test_compiled_autograd.py", line 1447, in fn
2024-07-29T21:37:39.7001053Z     loss.backward()
2024-07-29T21:37:39.7001638Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/_tensor.py", line 522, in backward
2024-07-29T21:37:39.7002209Z     torch.autograd.backward(
2024-07-29T21:37:39.7002784Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/autograd/__init__.py", line 346, in backward
2024-07-29T21:37:39.7003394Z     _engine_run_backward(
2024-07-29T21:37:39.7004005Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/autograd/graph.py", line 806, in _engine_run_backward
2024-07-29T21:37:39.7004892Z     return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
2024-07-29T21:37:39.7005889Z RuntimeError: compiled_args not implemented for non-traceable node: torch::autograd::CppNode<CustomOpAutogradFunction>
2024-07-29T21:37:39.7006430Z 
```

File: test/inductor/test_compiled_autograd.py [hunk 2]
```diff
@@ -2231,6 +2231,13 @@ def wrap_test_class(orig_cls):
 )
 
 # Bugs needing investigation:
+skipped_tests = {
+    # These test unconventional usage of saved tensor hooks do not leak or crash
+    # Running these tests succeed, but somehow cause other tests to fail
+    # "test_saved_tensor_hooks_extra_exit_during_bw_no_crash",
+    # "test_saved_tensor_hooks_extra_enter_during_bw_no_leak",
+}
+
```
Comment (soulitzer) [inline test/inductor/test_compiled_autograd.py:2240]: @xmfan The new tests in this PR causes a compiled_autograd test to break in a way that I don't understand.  https://github.com/pytorch/pytorch/pull/131700/files#r1698837389 how do you feel about skipping for now?

Comment (soulitzer) [reply test/inductor/test_compiled_autograd.py:2240]: I think I will skip for now. The PR should not break compiled autograd, it's just these tests are doing very weird things.

Comment (xmfan) [reply test/inductor/test_compiled_autograd.py:2240]: were they flaky? seems to pass in https://github.com/pytorch/pytorch/pull/134163

Comment (soulitzer) [reply test/inductor/test_compiled_autograd.py:2240]: That's interesting, they weren't flaky when I observed the failures, but fwiw the only way I could reproduce the failure is in CI and it was consistent the couple times it ran

**Pull request closed without merging**